polkadot/xcm/xcm-builder/src/transactional.rs

```diff
@@ -0,0 +1,39 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use frame_support::storage::{with_transaction_unchecked, TransactionOutcome};
+use xcm::latest::prelude::*;
+use xcm_executor::traits::ProcessTransaction;
+
+/// Transactional processor implementation using frame transactional layers.
+pub struct FrameTransactionalProcessor;
+impl ProcessTransaction for FrameTransactionalProcessor {
+	const IS_TRANSACTIONAL: bool = true;
+
+	fn process<F>(f: F) -> Result<(), XcmError>
+	where
+		F: FnOnce() -> Result<(), XcmError>,
+	{
+		// Is safe to use because of the `MAX_XCM_DECODE_DEPTH`.
+		with_transaction_unchecked(|| -> TransactionOutcome<Result<(), XcmError>> {
```

Comment:
Is this guaranteed to be used top-level? There could be a lot of `Utility::batch` nested on top of the XCM executor, or not?

Reply:
I think you might be right. If we nest 256 batch calls with the last call to the XCM executor, the executor could go over the 256 limit. I think in the current situation it would only be one extra layer, as I think we do not use any transactionality on XCM nesting instructions. I can change the implementation again to use with_transaction. The implementation is quite ugly as with_transaction expects a `From<DispatchError>`  while the closure returns a potential XcmError that does not implement `From<DispatchError>`.

Reply:
>I think in the current situation it would only be one extra layer, as I think we do not use any transactionality on XCM nesting instructions.

But XCM is possibly re-entrant AFAIK. One XCM can dispatch a RuntimeCalll which re-enters the XCM executor, or not?

Reply:
In one of my last commits, I updated it to `with_transaction`